subkey/src/main.rs

```diff
@@ -91,216 +108,366 @@ impl Crypto for Sr25519 {
+		.expect("parameter is required; thus it can't be None; qed");
+	str::parse::<T>(str_value).expect("Invalid 'nonce' parameter; expecting an integer.")
+}
+
+fn read_genesis_hash(matches: &ArgMatches) -> H256 {
+	let genesis_hash: Hash = match matches.value_of("genesis").unwrap_or("alex") {
+		"elm" => hex!["10c08714a10c7da78f40a60f6f732cf0dba97acfb5e2035445b032386157d5c3"].into(),
+		"alex" => hex!["dcd1346701ca8396496e52aa2785b1748deb6db09551b72159dcb3e08991025b"].into(),
+		h => hex::decode(h)
+			.ok()
+			.and_then(|x| Decode::decode(&mut &x[..]).ok())
+			.expect("Invalid genesis hash or unrecognised chain identifier"),
+	};
+	println!(
+		"Using a genesis hash of {}",
+		HexDisplay::from(&genesis_hash.as_ref())
+	);
+	genesis_hash
+}
+
+fn read_signature<C: Crypto>(matches: &ArgMatches) -> SignatureOf<C>
+where
+	SignatureOf<C>: SignatureT,
+	PublicOf<C>: PublicT,
+{
+	let sig_data = matches
+		.value_of("sig")
+		.expect("signature parameter is required; thus it can't be None; qed");
+	let mut signature = <<C as Crypto>::Pair as Pair>::Signature::default();
+	let sig_data = hex::decode(sig_data).expect("signature is invalid hex");
+	if sig_data.len() != signature.as_ref().len() {
+		panic!(
+			"signature has an invalid length. read {} bytes, expected {} bytes",
+			sig_data.len(),
+			signature.as_ref().len(),
+		);
+	}
+	signature.as_mut().copy_from_slice(&sig_data);
+	signature
+}
+
+fn read_public_key<C: Crypto>(matched_uri: Option<&str>, password: Option<&str>) -> PublicOf<C>
+where
+	SignatureOf<C>: SignatureT,
+	PublicOf<C>: PublicT,
+{
+	let uri = matched_uri.expect("parameter is required; thus it can't be None; qed");
+	let uri = if uri.starts_with("0x") {
+		&uri[2..]
 	} else {
-		execute::<Sr25519>(matches)
+		uri
+	};
+	if let Ok(pubkey_vec) = hex::decode(uri) {
+		<C as Crypto>::Public::from_slice(pubkey_vec.as_slice())
+	} else {
+		<C as Crypto>::Pair::from_string(uri, password)
+			.ok()
+			.map(|p| p.public())
+			.expect("Invalid URI; expecting either a secret URI or a public URI.")
 	}
 }
 
-fn print_usage(matches: &clap::ArgMatches) {
+fn read_pair<C: Crypto>(
+	matched_suri: Option<&str>,
+	password: Option<&str>,
+) -> <C as Crypto>::Pair
+where
+	SignatureOf<C>: SignatureT,
+	PublicOf<C>: PublicT,
+{
+	let suri = matched_suri.expect("parameter is required; thus it can't be None; qed");
+	C::pair_from_suri(suri, password)
+}
+
+fn format_signature<C: Crypto>(signature: &SignatureOf<C>) -> String {
+	format!("{}", hex::encode(signature))
+}
+
+fn format_seed<C: Crypto>(seed: SeedOf<C>) -> String {
+	format!("0x{}", HexDisplay::from(&seed.as_ref()))
+}
+
+fn format_public_key<C: Crypto>(public_key: PublicOf<C>) -> String {
+	format!("0x{}", HexDisplay::from(&public_key.as_ref()))
+}
+
+fn create_extrinsic(
+	function: Call,
+	index: Index,
+	signer: <Sr25519 as Crypto>::Pair,
+	genesis_hash: H256,
+) -> UncheckedExtrinsic {
+	let extra = |i: Index, f: Balance| {
+		(
+			system::CheckVersion::<Runtime>::new(),
```

Comment:
at some point we should: 
- allow subkey to create transactions with different extension list (for different potential chains)
- add integration tests for them.

For now, I think it is worthwhile to make sure that the extension format here is in sync with `node/runtime`.

Reply:
TBH, I couldn't find a satisfactory way to test this function (and the respective commands). Also, I  don't know if people still use them.